src/README.md

```diff
@@ -0,0 +1,26 @@
+This README serves as developer documentation for CSV.jl internals. It doesn't pretend to be comprehensive, but was created with the aim to explain both the overall strategies CSV.jl uses in parsing delimited files as well as the trickiest mechanics and some of the reasoning that went into the architecture.
+
+# CSV.File
+
+Let's go over details specific to `CSV.File`.
+
+## The `pool` keyword argument
+
+CSV.jl provides a native integration with the [PooledArrays.jl](https://github.com/JuliaData/PooledArrays.jl/) package, which provides an array storage optimization by having a (hopefully) small pool of "exspensive" (big or heap-allocated, or whatever) values, along with a memory-efficient integer array of "refs" where each ref maps to one of the values in the pool. This is sometimes referred to as a "dictionary encoding" in various data formats. As an example, if you have a column with 1,000,000 elements, but only 10 unique string values, you can have a `Vector{String}` pool to store the 10 unique strings and give each a unique `UInt32` value, and a `Vector{UInt32}` "ref array" for the million elements, where each element just indexes into the pool to get the actual string value.
+
+By providing the `pool` keyword argument, users can control how this optimization will be applied to individual columns, or to all columns of the delimted text being read.
+
+Valid inputs for `pool` include:
+  * A `Bool`, `true` or `false`, which will apply to _all_ columns parsed; they either will _all_ be pooled, or _all_ not pooled
```

Comment:
do we think this is useful in practice? E.g. under these rules `Bool` column would be pooled.

Reply:
Yeah, perhaps we should still limit this case to `stringtype` columns; that would let us get rid of the `NaN` special-case, because we could change the default back to `pool=0.1` (or whatever threshold value we want).

Reply:
I think pooling `<:Number` does not make much sense. I would restrict the default to `stringtype` columns. @nalimilan - any opinion here? Just to be sure: what column `eltype`s CSV.jl by default can create? I understand that apart from numbers and strings it is only dates. Right?

Reply:
Int64, Float64, Bool, Date, DateTime, Time, and `stringtype`. But we do allow passing custom types. I needed to generalize the pooling code to eventually support pooling short strings, and allowing any type naturally comes out of that. I agree it doesn't make sense to pool Bool and probably Float64, but there may be a use-case for Int64. I think it would be a good default to still only pool strings by default, but allow user to pool custom type/Int64 if requested.

Reply:
>  I think it would be a good default to still only pool strings by default, but allow user to pool custom type/Int64 if requested.

Agreed